Design/DesignDoc-Least-ConnectionsTCPLB

```diff
@@ -0,0 +1,181 @@
+```
+# Generate CA cert
+openssl genrsa -aes256 -out ca/ca.key 4096 
+openssl req -new -x509 -sha256 -days 20 -key ca/ca.key -out ca/ca.crt
+
+# Generate CSR for server
+openssl genrsa -out server/localhost.key 2048
+openssl req -new -key server/localhost.key -sha256 -out server/localhost.csr
+# Validate CSR
+openssl x509 -req -days 365 -sha256 -in server/localhost.csr -CA ca/ca.crt -CAkey ca/ca.key -set_serial 1 -out server/localhost.crt
+
+# Create client CRT
+openssl genrsa -out client/client1.key 2048
+openssl req -new -key client/client1.key -out client/client1.csr
+# Validate CSR
+openssl x509 -req -days 365 -sha256 -in client/client1.csr -CA ca/ca.crt -CAkey ca/ca.key -set_serial 2 -out client/client1.crt
+```
+
+#### TLS versions
+We will support TLS 1.2 and 1.3 for the communication between clients and server. It removes a lot of unsecure or deprecated TLS features and provides better security. I found out that an estimate of 20% of corporate internet traffic is using TLS 1.3 in 2023 and adding TLS 1.2 will cover the most secure enough clients possible. 
+
+#### Cipher suites
+A cipher suite consists of a key exchange algorithm, an authentication algorithm, a bulk encryption algorithm, and a message authentication algorithm.
+
+The following cipher suites can be used from must secure to less secure: 
+1. ECDHE-ECDSA-AES256-GCM-SHA384
+2. ECDHE-RSA-AES256-GCM-SHA384
+3. ECDHE-ECDSA-CHACHA20-POLY1305
+4. ECDHE-RSA-CHACHA20-POLY1305
+5. ECDHE-ECDSA-AES128-GCM-SHA256
+6. ECDHE-RSA-AES128-GCM-SHA256
+
+
+#### Configuration
+We can use [Mozilla SSL Config Generator](https://ssl-config.mozilla.org/#server=go&version=1.21&config=intermediate&guideline=5.7) with the intermediate configuration and extract the Go TLS configuration as follows
+```
+cfg := &tls.Config{
+    MinVersion: tls.VersionTLS12,
+    CipherSuites: []uint16{
+        tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
+        tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
+        tls.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
+        tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
+        tls.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,
+        tls.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,
+    },
+}
+```
+### Authorization scheme
+#### Principle
+A list is created in the server with the upstreams associated to each client. This list is passed to the forward method who will verify if the upstream selected is in the list passed for the client. The upstreams list and the client's accessible upstream list are created by the same layer so if the list is empty we do not forward the request and close the connection.
+
+For example the list will be passed like that:
+```go
+upstreams := []string{"localhost:9001", "localhost:9002"}
+go forwarder.Forward(clientConn, upstreams)
+```
+
+The way to determine the client is the subject.CommonName field of the certificate. The certificate needs to have that information correctly filled so that we can identify it and authorize it.
```

Comment:
Can you give an example of how the `CommonName` is mapped to the backend?

Reply:
I will add this detail in the Certs generation part